enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+
+    # Optional directory containing the images. When this is used the machine config
+    # operator will expect it to contain a dump of a registry server, will start a
+    # temporary instance of the registry server and will pull the pinned images
+    # from it.
+    #
+    # If this not specified the images will be pulled from their origin or from the
+    # mirrors configured in the cluster.
+    imagesDirectory: /var/lib/upgrade/4.13.7-x86_64
+
+status:
+  containerRuntimeConfig:
+
+    # List of effectively pre-loaded and pinned images. It will be populated by
+    # the machine config operator when the images have been successfully pre-oaded
+    # and pinned.
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+   ...
+```
+
+### Implementation Details/Notes/Constraints
+
+The proposed solution is based on pre-loading and pinning all required images
+in all the nodes of the cluster before starting the upgrade, and ensuring that
+no component requires access to a registry server during the upgrade. For this
+to work the following changes are required:
+
+1. No OpenShift component used during the upgrade should use the `Always` pull
+policy, as that forces the kubelet and CRI-O to try to contact the registry
+server even if the image is already available.
+
+1. No OpenShift component should garbage collect the images required for the
+upgrade. This is typically started by the kubelet instructing CRI-O to remove
+images.
+
+1. No OpenShift component should explicitly try to contact the registry server
+without a fallback alternative.
+
+1. The machine config operator needs to support image pinning, pre-loading and
+reloading of the CRI-O service.
+
+1. The cluster version operator needs to orchestrate the upgrade process.
+
+1. The engineer that prepares the upgrade bundle needs a `oc adm upgrade create
+bundle` tool to create it.
+
+1. The technician that performs the upgrade needs documentation explaining how
+to use the upgrade bundle.
+
+#### Don't use the `Always` pull policy during the upgrade
+
+Some OCP core components currently use the `Always` image pull policy during the
+upgrade. As a result, the kubelet and CRI-O will try to contact the registry
+server, even if the image is already available in the local storage of the
+cluster. This blocks the upgrade.
+
+The catalog operator uses the `Always` pull policy to pull catalog images. It
+does so in order refresh catalog images that are specified with a tag. But it
+also does it when it pulls catalog images that are specified with a digest. That
+should be changed to use the `IfNotPresent` pull policy for catalog images that
+are specified by digest.
```

Comment:
@donpenney that's what you mentioned, right?

Reply:
No, I was referring to MCO's behaviour during an upgrade. The MCO has (or had) code that runs during the upgrade that is pulling manifests from the image registry, regardless of whether the image is available locally. It looks like @cgwalters has now removed this code in 4.14, so that wouldn't be an issue moving forward:
https://github.com/openshift/machine-config-operator/commit/85b297c1a063a7e130e82afb2fd2230ffd5c5fcc

Reply:
There's still more registry fetching code that needs to be removed, see https://github.com/openshift/machine-config-operator/pull/3821

Reply:
I agree. If we want to limit this enhancement in scope, it should cover how we upgrade the cluster without registry, rather than how we operate OCP without a registry.

Reply:
That depends on the approach that was used to install the cluster. In the appliance approach there is no netboot, but a recovery partition and boot entry that the customer can use to completely reinstall the cluster. That recovery partition is preloaded before the cluster is shipped to the site. But if a drive fails it usually means that the complete node failed (and even the complete cluster, if it is multi-node). In that case the customer either replaces the drive with a new one (containing that recovery partition, like the original) or else replaces the complete cluster enclosure.